### ISSUE TITLE
$mobile_device not working anymore

### ISSUE THREAD

**Issue by Zengraph** [thirtybees/thirtybees#1030]
the mobile detection is not working anymore as it always have the value "false", something is broken somewhere and should be updated, very usefull to know if it's a mobile/tablet or desktop in order to make the best theme adaptation for each, that concern JS scripts as well that sometimes don't work same on mobile or desktop, like testing screen resizing or other functions.

**Comment by Traumflug** [comment]
Actually, the plan is to remove this detection for the reason you describe: it's unreliable. Depending on screen sizes is much more reliable. Bootstrap uses this natively and JS can figure this easily as well.

**Comment by Zengraph** [comment]
"Mobile Detect is a lightweight PHP class for detecting mobile devices (including tablets). It uses the User-Agent string combined with specific HTTP headers to detect the mobile environment." => it is working good, even it's not bullet proof but it can help you to determine what to do for each case.

Css and JS can handle the sizes of the screen and the resize of it but on mobile sometimes this JS resizing check is making trouble, so I would like to turn it off if the screen width is small AND Http headers is android or IOS

**Comment by Dh42** [comment]
I don't think it would be good to remove this at this point either. Its a great feature that can help save loading time for mobile sites, which are the most popular sites these days.

**Comment by Zengraph** [comment]
Yes dh42, that is why I'd like to optimize it by recovering the $is_mobile value on front office in .tpl and JS, that does not work anymore. Plz note that module configuration for tablets/mobile Activation/Loading is still working good.. just a small fix maybe.

**Comment by Traumflug** [comment]
`$is_mobile = window.innerWidth < 992` ?

**Comment by Dh42** [comment]
Is that all the core uses to determine mobile?

**Comment by Zengraph** [comment]
It uses Mobile Detect php code, based on http agent, that's the best avalaible today, even it's not perfect it does the job

**Comment by mke-tb** [comment]
@Zengraph What do you need exactly, having access to `$is_mobile` in `tpl` files? Can you provide some information to reproduce the issue when mobile detection works incorrectly?

**Comment by mke-tb** [comment]
> `$is_mobile = window.innerWidth < 992` ?

@Traumflug What do you mean by that?

**Comment by Traumflug** [comment]
>> $is_mobile = window.innerWidth < 992 ?
>
> @Traumflug What do you mean by that?

Well, major distinction between a mobile and a not mobile device is width of the screen/window. Testing for that width is a simple method to find out which device the page gets rendered on.

More complex methods, like looking at the device or browser name are a bit questionable, IMHO. Because screen layouts are made for certain screen widths (Bootstrap entirely depends on it) and things mess up easily if a page on a 800px wide mobile device renders differently than in a 800px wide window on a desktop screen.

These days we even have [hybrid devices](https://www.techradar.com/news/mobile-computing/laptops/best-2-in-1-laptops-top-5-hybrid-laptops-reviewed-1258451); devices which can be used as mobile as well as laptops. There's no way to distinguish whether such a hybrid device is currently used mobile or stationary, OS is the same, browser is the same, hardware is the same. Still rendering should happen suitable for both usage modes.

**Comment by Traumflug** [comment]
To me it looks like fixing the underlying issue is the better approach. @Zengraph wrote above:

> Css and JS can handle the sizes of the screen and the resize of it but on mobile sometimes this JS resizing check is making trouble

**Comment by Zengraph** [comment]
Well, let me explain where my problem is : when user resize their window from full screen to mobile and do that again and again... that causes JS bugs many times on many websites, depending on browser, config, position of the stars in the sky.. who knows :) Otherwise if a user is using mobile or tablet.. he can just change position of the screen but never reach from 600px to 2400px large and If it's know that it's not a desktop then many JS will not be used like resizing check for exemple, used for exemple in a left column that never scroll and stay fixed in desktop, not in tablet/mobile, but if a desktop user is opening this website on a small window, if the resizing is not checked in JS then the column will never be fixed on desktop, but this resizing check is not necessary on small screens.

We can also check the Pixel ratio to guess if the screen is of a mobile or not.

One more exemple : the home page slider module can be inactive on mobile/tablet and that works on module config, That's great because instead you can just show a good pic with the proper ratio instead of loading 3 or 4 huge images in UHD.
.

**Comment by Dh42** [comment]
The main problem I see with using a js or css method is that it defeats the point. One thing I like to do with sites is use that feature to create a lighter mobile version. Like say on a desktop site the home page consists of the slider, a new products block, a feature products block, and a top sellers block. On mobile have the new products disabled and the slider as well. Disabling those with javascript or css will still load those assets. Since mobile devices are slower, you will still get the lag because the javascript is still running and you will also still get the larger page size.

**Comment by Traumflug** [comment]
True, this topic "light pages" needs a solution.

**Comment by Dh42** [comment]
Is something wrong with the current library we use? https://github.com/serbanghita/Mobile-Detect

**Comment by Zengraph** [comment]
library seems ok because it works on modules setting where you can activate them for each device (mobile / tablet /desktop), but the functionallity don't work anymore on smarty themes files: TPL.

**Comment by mke-tb** [comment]
From what I understand @dh42 is more in favor of using mobile detection on the PHP side, not JS. I agree with that, mobile compability shouldn't go beyond responsive CSS and maybe whatever JS code that's there with Bootstrap libraries.

If we maintain the PHP part, anybody who wants can easily define it in his JS and use.

**Comment by Dh42** [comment]
I think what @Zengraph  is saying is that a variable used to be exposed on a template level that would return true/false based on the device. This can still be done really, but would likely cause problems in the page cache.

**Comment by mke-tb** [comment]
What kind of problems do you anticipate?

**Comment by Dh42** [comment]
If we are using the full page cache it would generate pages with parts missing, so we would need 3 sets of full page cache files, mobile, tablet, and desktop. You might ask @getdatakick  what his thoughts are, I know he has worked with the cache more than any of us.

**Comment by getdatakick** [comment]
That's correct. At the moment, FPC has problems with different server-side handling of request for mobile/tablet/desktop. Because we can have modules enabled on some devices only, the device type should be part of [page cache key](https://github.com/thirtybees/thirtybees/blob/136f43814634ed3a0e3da33bc19a9e56231e4ed6/classes/PageCacheKey.php#L29-L37). It is not.

Adding new variable to smarty template will not make it worse. It's already bad.

**Comment by kristho91** [comment]
Ultimately, I think that mobile-detection is a frontend responsibility (handled in CSS and JS).
Also, I think that from a SEO standpoint it's a bad idea to remove content when on a mobile device, just for making the site "lighter". Google will rank your page based on the mobile version of the site.

**Comment by Dh42** [comment]
I disagree. The idea behind the mobile detection is useful, it just is not used a lot of time. There is no issue at all in showing mobile users one thing and desktop users another thing, almost every major ecommerce site uses that strategy from a server side integration. 


Take Walmart, who currently spends more on e-commerce technology than any other company in the world. They have both a mobile and a desktop site that differ based on browser / mobile detection that happens server side. Its generally smaller things you might not notice, but they do it. An example from their current home page is on desktop you are shown 8 categories in their category section on the home page. On mobile you are only shown 4. The area under the banner you are shown 5 on desktop, while on mobile it shrinks to 2 columns of 2 items, so 4. Both of these changes are done server side, so they are not manipulating with css. 


The idea behind using css and javascript is two fold bad. Both of the examples I gave with Walmart could be done with css, just using a media query and a display:none. But that will hurt your pagespeed since the data is still loaded. You could use a more advanced js method to add elements based on the screen size, but that has its drawbacks as well. Notably that it will descrease your page speed score because it will increase your time to interactive.

**Comment by getdatakick** [comment]
There are two reasons why this server-based feature is needed

### Performance

with js/css only solution, server still needs to load the module, run bunch of sql queries to retrive data, and generate html content from it. Browser then has to  parse much bigger dom tree, load more javascript, process more css rules, etc. 

Basically, both server and browser have to work very, very hard only to achieve the same result as if the module wasn't installed at all. Performance waste. 

### UX experience

With server-side solution, merchants can simply disable unwanted modules in module list. With client-side solution, this would not work. Merchants would have to understand how CSS works in order to hide unwanted content from the page. They don't want that, they don't know how to do that.
